File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Dictionary to store the most recent index of each character
        char_map = {}
        max_length = 0
        start = 0

        for end in range(len(s)):
            if s[end] in char_map:
                # If the character is already in the dictionary, move the start pointer
                # to the right of the last occurrence of s[end] + 1
                start = max(char_map[s[end]] + 1, start)

            # Update the last seen position of the character
            char_map[s[end]] = end

            # Update the maximum length found so far
            max_length = max(max_length, end - start + 1)

        return max_length

# Example usage:
solution = Solution()

s1 = "abcabcbb"
print(f"Input: s = \"{s1}\"")
print("Output:", solution.lengthOfLongestSubstring(s1))  # Output: 3

s2 = "bbbbb"
print(f"Input: s = \"{s2}\"")
print("Output:", solution.lengthOfLongestSubstring(s2))  # Output: 1

s3 = "pwwkew"
print(f"Input: s = \"{s3}\"")
print("Output:", solution.lengthOfLongestSubstring(s3))  # Output: 3

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        // Map para armazenar o índice mais recente de cada caractere
        Map<Character, Integer> charMap = new HashMap<>();
        int maxLength = 0;
        int start = 0;

        for (int end = 0; end < s.length(); end++) {
            char currentChar = s.charAt(end);

            if (charMap.containsKey(currentChar)) {
                // Se o caractere já está no mapa, move o ponteiro de início
                // para a direita da última ocorrência de currentChar + 1
                start = Math.max(charMap.get(currentChar) + 1, start);
            }

            // Atualiza o índice mais recente do caractere
            charMap.put(currentChar, end);

            // Atualiza o comprimento máximo encontrado até agora
            maxLength = Math.max(maxLength, end - start + 1);
        }

        return maxLength;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        String s1 = "abcabcbb";
        System.out.println("Input: s = \"" + s1 + "\"");
        System.out.println("Output: " + solution.lengthOfLongestSubstring(s1));  // Output: 3

        String s2 = "bbbbb";
        System.out.println("Input: s = \"" + s2 + "\"");
        System.out.println("Output: " + solution.lengthOfLongestSubstring(s2));  // Output: 1

        String s3 = "pwwkew";
        System.out.println("Input: s = \"" + s3 + "\"");
        System.out.println("Output: " + solution.lengthOfLongestSubstring(s3));  // Output: 3
    }
}
